vcd/resource_vcd_vapp_vm_tools.go

```diff
@@ -695,7 +695,7 @@ func updateTemplateInternalDisks(d *schema.ResourceData, meta interface{}, vm go
 		}
 
 		// Update details of internal disk for disk existing in template
-		if value, ok := internalDiskProvidedConfig["iops"]; ok {
+		if value, ok := internalDiskProvidedConfig["iops"]; ok && diskCreatedByTemplate.IopsAllocation != nil {
```

Comment:
I have a few thoughts for discussion

The way it checks now is - if VM that was spun from template has `diskCreatedByTemplate.IopsAllocation` defined - we're setting the `iops` field. Otherwise we ignore.

On one hand - it validates if the customer is able to possibly read (right `View Disk IOPS`) the disk structure. However we'd still not be sure if he has right `Edit Disk IOPS` and in the end it still might cause panic.

Additionally, given how complicated the APIs may be in some cases - it could be that the VM template simply didn't have this `diskCreatedByTemplate.IopsAllocation` structure and VCD doesn't return it? If such case happens (I don't know if it can happen) - we might start ignoring `iops` even if the user set it that could lead to another bug.

The issue that user has (#1262) - he is not setting the `iops` field (at least in his sample snippet). But the condition `if value, ok := internalDiskProvidedConfig["iops"]; ok {` looks to be always `true` because the field is `schema.TypeInt` (which is `0`, when unset). 

My thought is - should we additionally validate another condition - that the user has set the field `iops` explicitly to be non zero (`value.(int) != 0`). If so - even if he didn't have rights (`diskCreatedByTemplate.IopsAllocation == nil`, to be precise)  - we could initialise the structure and send it. We'd prevent panic and the user would get API error that he doesn't have rights. He would just then know that he shouldn't set `iops` field.

_My thoughts in code (unvalidated)_

``` go
if value := internalDiskProvidedConfig["iops"]; value.(int) != 0 || diskCreatedByTemplate.IopsAllocation != nil {
	if diskCreatedByTemplate.IopsAllocation == nil {
		diskCreatedByTemplate.IopsAllocation = &types.IopsResource{}
	}
	iops := int64(value.(int))
	diskCreatedByTemplate.IopsAllocation.Reservation = iops
}
```

Reply:
Actually that's a good point, I'll reiterate the implementation with this in mind and push an update.

Reply:
Changed. Test passed! I'll run `vm` tag tests just in case, before merging.